### ISSUE TITLE
Generated app should check if foreman is installed

### ISSUE THREAD

**Issue by grepsedawk** [luckyframework/lucky_cli#714]
In the generated app, it should check if foreman is installed in https://github.com/luckyframework/lucky_cli/blob/master/src/web_app_skeleton/script/system_check.ecr

If foreman isn't installed, `lucky dev` will fail to run

**Comment by jwoertink** [comment]
The CLI actually already checks for this.
https://github.com/luckyframework/lucky_cli/blob/076d93e34fa91eb4bedbb4a5dd6eaca08fdc3cfc/src/lucky_cli/process_runner.cr#L75
https://github.com/luckyframework/lucky_cli/blob/076d93e34fa91eb4bedbb4a5dd6eaca08fdc3cfc/src/lucky_cli/process_runner.cr#L22

If you only have foreman installed, it'll use that one, but it's last on the list. So if any of the other ones are installed too, then foreman is ignored.

Now, this probably won't matter soon anyway with https://github.com/luckyframework/lucky_cli/pull/710 We want to get away from 3rd party tools where we can because we've noticed that they increase barrier to entry. So with a process manager built-in, then it's 1 less thing for a newcomer to need to get running!

**Comment by grepsedawk** [comment]
Ooh I see what is happening here!
Most of the major Ruby VM's utilize "shims" for globally installed versions
I had a `foreman` shim, but it actually wasn't installed (IE `foreman -v` would fail), but there was still an executable matching `foreman`

Testing against `foreman -v` would fix this, but it would also take more time -- so I think it would either need to be done in the `script/system_check` template (only invoked by `script/setup`), or this be deemed enough of an edge case that this issue can be closed

_**tl;dr - This is an edge case when using a Ruby version manager that uses a shimmed pattern, feel free to close this issue**_

**Comment by jwoertink** [comment]
Oh, so you're saying `Process.find_executable("foreman")` fails, but `Process.find_executable("foreman -v")` works? 🤔  Is that a Crystal bug? My understanding is the shim is just a symlink. If `which foreman` works, then `Process.find_executable("foreman")` should also work... I'm not going to close this just yet if there's a bug with how we pick the runner. But once nox is tested out, then it'll be more of just docs on how to use other process runners.

**Comment by grepsedawk** [comment]
`!!Process.find_executable("foreman") #=> true`
`system("foreman -v")`

results in non-0 exit code, because the existing shim points to a real file, but the shim itself fails to execute because it's not installed on current ruby

**Comment by jwoertink** [comment]
ooohhhhh 😂 I was confused for a min there, but I see now.  I have foreman installed the same way.

```
Development/crystal/sandbox via 🔮 v1.2.2
❯ cat test.cr
pp Process.find_executable("foreman")

Development/crystal/sandbox via 🔮 v1.2.2
❯ crystal test.cr
"/Users/jeremywoertink/.asdf/shims/foreman"

Development/crystal/sandbox via 🔮 v1.2.2
❯ which foreman
/Users/jeremywoertink/.asdf/shims/foreman

Development/crystal/sandbox via 🔮 v1.2.2
❯ foreman -v
No preset version installed for command foreman
Please install a version by running one of the following:

asdf install ruby 2.7.2

Development/crystal/sandbox via 🔮 v1.2.2
❯ bundle exec foreman start
Could not locate Gemfile or .bundle/ directory
```

Ok, so I think I see the issue here. It's probably moot anyway, but basically we're saying "foreman exists, but you can't use it" since you'd normally do `bundle exec foreman start` to utilize the shim. It's definitely a sort of edge case, but also a bug in the current versions of Lucky.

For anyone else looking to use foreman, you'll probably just have to use `foreman start -f Procfile.dev` as opposed to `lucky dev` for now.

**Comment by grepsedawk** [comment]
Actually it can be resolved by doing `gem install foreman`, as `.asdf/shims/foreman` is intended to be in included in the path, the shim just fails in a slightly awkward way.
This should express this a lot better:

```plaintext
➜  hello-lucky git:(main) ✗ asdf global ruby 3.1.0
➜  hello-lucky git:(main) ✗ ruby -v
ruby 3.1.0p0 (2021-12-25 revision fb4df44d16) [x86_64-linux]
➜  hello-lucky git:(main) ✗ lucky dev
No preset version installed for command foreman
Please install a version by running one of the following:

asdf install ruby 3.1.0

or add one of the following versions in your config file at /home/alex/.tool-versions
ruby 2.7.2
ruby 3.0.2
➜  hello-lucky git:(main) ✗ gem install foreman
Fetching foreman-0.87.2.gem
Successfully installed foreman-0.87.2
Parsing documentation for foreman-0.87.2
Installing ri documentation for foreman-0.87.2
Done installing documentation for foreman after 0 seconds
1 gem installed
➜  hello-lucky git:(main) ✗ lucky dev
13:36:11 system_check.1 | started with pid 1588847
13:36:11 web.1          | started with pid 1588848
13:36:11 assets.1       | started with pid 1588849
13:36:11 web.1          | Beginning to watch your project
```

I was thinking if we add a `foreman -v` check to , we can alert the user before they try to start the app that foreman isn't working, during the setup phase so it doesn't slow down actual CLI operations

I haven't checked other ruby VM's, but I'm guessing the error would be around this obscure. Definitely not lucky's fault here.

**Comment by matthewmcgarvey** [comment]
@jwoertink mentioned it in passing, but to make it clear for anyone looking at this issue...

While it is an issue, this is not something you should spend your time working on. In the next version of Lucky, we're probably going to be removing all these ProcessRunners (including foreman's) and replacing them with one written in Crystal. See https://github.com/luckyframework/lucky_cli/pull/710

On the other hand, it would be a good idea to see if there's any other code that does these kinds of checks. I can't think of any off the top of my head.
